Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp

```diff
@@ -88,13 +89,41 @@ static CBORValue convertDescriptorToCBOR(const PublicKeyCredentialDescriptor& de
     return CBORValue(WTFMove(cborDescriptorMap));
 }
 
-Vector<uint8_t> encodeMakeCredentialRequestAsCBOR(const Vector<uint8_t>& hash, const PublicKeyCredentialCreationOptions& options, UVAvailability uvCapability, AuthenticatorSupportedOptions::ResidentKeyAvailability residentKeyAvailability, const Vector<String>& authenticatorSupportedExtensions, std::optional<PinParameters> pin)
+static Vector<PublicKeyCredentialParameters> trimmedParameters(const Vector<PublicKeyCredentialParameters>& parameters, const std::optional<Vector<WebCore::PublicKeyCredentialParameters>>& authenticatorSupportedParameters)
+{
+    HashSet<int64_t> authenticatorSupportedAlgorithms;
+    if (authenticatorSupportedParameters) {
+        for (auto& parameters : *authenticatorSupportedParameters) {
+            if (parameters.type == PublicKeyCredentialType::PublicKey)
+                authenticatorSupportedAlgorithms.add(parameters.alg);
+        }
+    }
+
+    for (auto& parameter : parameters) {
+        if (parameter.type != PublicKeyCredentialType::PublicKey)
+            continue;
+        if (authenticatorSupportedAlgorithms.contains(parameter.alg))
+            return { parameter };
+        // Support for ES256 required by U2F backwards compatibility.
+        // https://fidoalliance.org/specs/fido-v2.0-id-20180227/fido-client-to-authenticator-protocol-v2.0-id-20180227.html#u2f-authenticatorMakeCredential-interoperability
+        if (parameter.alg == COSE::ES256)
+            return { parameter };
+    }
+
+    // We know the algorithms the authenticator supports and none of those were requested.
+    if (authenticatorSupportedAlgorithms.size())
+        return { parameters.first() };
```

Comment:
Why is it better to return just the first unsupported algorithm, rather than the entire set? I don't really understand why this is better for a case where we can't use any of the parameters.

Reply:
The issue is that some websites will ask for every last algorithm they support in the list they provide, which can cause the message sent for makeCredential to become too large to be sent over a transport like NFC. The idea with this patch is that we know the processing steps the authenticator will take from the CTAP spec and the authenticator already reports what algorithms it supports in its getInfo response.

Using that we can prune the list to avoid hitting maxMsgSize whenever the RP sends a long list.

For example, apparently one RP sends a request like:
`{ "attestation": "direct", "authenticatorSelection": { "residentKey": "discouraged", "userVerification": "preferred" }, "challenge": "a0oPIQ4k5mF15rK1WKnUxqGsRYvsC-0tHf3er8i0LZHTv8f4KGUMRQ1ys- aJiFg0_OuIPhKA8ecaRuUmfHN3EQ", "excludeCredentials": [ {"id": "iNRyobHFV_khWBgdUXpkeNxVFjRh2MIZI7YSJt4qBDTsK1kwQ8H6DmqixmZM2jRIbhXpbSV_UN_9GK4PcuMYv4EMReqzRhIPAgggCK21ouy4fHE HUFHHAYdMTC6d8Q", "transports": [ "usb", "nfc", "ble", "hybrid", "internal" ], "type": "public-key" }, { "id": "vxwsDOgJ08syZrS13YAEXS7UV51B0rb8гBTm92NJvWaxMEQOUkc2L_Nvs Se6Cprd@VG5Wt-0WUcuDx5xdomaIA", "transports": [ "usb", "nfc", "ble", "hybrid", "internal" ], "type": "public-key" } ], "extensions": { "credProps": true }, "pubKeyCredParams": [ { "alg":-65535, "type": "public-key" }, { "alg":-257, "type": "public-key" }, { "alg": -258, "type": "public-key" }, { "alg": -259, "type": "public-key" }, { "alg": -37, "type": "public-key" }, { "alg": -38, "type": "public-key" }, { "alg":-39, "type": "public-key" }, { "alg": 7, "type": "public-key" }, { "alg":-35, "type": "public-key" }, { "alg": -36, "type": "public-key" }, { "alg": -8, "type": "public-key" }, { "alg": -43, "type": "public-key" } ], "rp": { "id": "...", "name": "..." }, "user": { "displayName": "Chrome V126.0.0.0 (Windows)", "id": "NDg2MTA4NDIz", "name": "[xxxxxx.user1@gmail.com](mailto:xxxxxx.user1@gmail.com)" } }`

Sending all those algorithms can easily exceed the maxMsgSize for NFC, especially because "public-key" is encoded as a string in the CTAP spec.

Reply:
It's not surprising that returning one response is a smaller message than all possible things. I'm just unclear why returning the first thing (highest preference) algorithm our authenticator supports when we don't match any of the things the website asked for is appropriate. Shouldn't we have some kind of error state or something?

Reply:
My logic here was just to maintain the same behavior wrt to the authenticator with the original list vs the truncated list. In both cases the authenticator will return CTAP2_ERR_UNSUPPORTED_ALGORITHM and our existing error handling code will handle that. I'd like to avoid creating two separate paths for this error to get handled.

You bring up a good point about the error handling here. Right now we actually do the wrong thing by not handling errors like `CTAP2_ERR_UNSUPPORTED_ALGORITHM`, `CTAP2_ERR_INVALID_OPTION `, `CTAP2_ERR_UNSUPPORTED_OPTION` returned from the authenticator.

These errors are returned before an authorization gesture is collected by the authenticator. It doesn't makes sense to report the error right after we get it back from the makeCredential request, because collecting an authorization gesture is the method used to select an authenticator to use for the ceremony.

For example, if the user has 3 security keys plugged in (or an unsupported security key plugged in, but another that is supported that they plug in during the ceremony,) it doesn't make sense to report the error whenever we get it back from makeCredential because we don't know if the authenticator that reported the error is the one the user wants to use for the ceremony. 

Instead we should note down we get this error from a particular authenticator, we should submit a dummy request to the authenticator to collect a tap. If they do tap the authenticator where we got this error, we can show a message that this authenticator cannot be used because it doesn't support an algorithm the RP wants.